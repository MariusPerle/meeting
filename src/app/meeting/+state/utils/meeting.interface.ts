export interface Record {
    id: string;
    name: string;
    date: Date;
    members: RecordMember[];
}

export interface RecordMember {
    id: string;
    time: number;
    note: string;
    attended: boolean;
    participated: boolean;
}

export function isMeeting(data: object): data is Record {
    return (
        'id' in data &&
        typeof data.id === 'string' &&
        'name' in data &&
        typeof data.name === 'string' &&
        'date' in data &&
        data.date instanceof Date &&
        !isNaN(data.date.getTime()) &&
        'members' in data &&
        Array.isArray(data.members) &&
        data.members.every(isMeetingMember)
    );
}

export function isMeetingMember(data: object): data is RecordMember {
    return (
        'id' in data &&
        typeof data.id === 'string' &&
        'time' in data &&
        typeof data.time === 'number' &&
        'note' in data &&
        typeof data.note === 'string' &&
        'attended' in data &&
        typeof data.attended === 'boolean' &&
        'participated' in data &&
        typeof data.participated === 'boolean'
    );
}
