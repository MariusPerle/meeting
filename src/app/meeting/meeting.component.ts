import { Component, inject, OnInit } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterOutlet } from '@angular/router';
import { AppService } from '../app.service';

@Component({
    imports: [CommonModule, RouterOutlet],
    templateUrl: './meeting.component.html',
    styleUrl: './meeting.component.scss',
})
export class MeetingComponent implements OnInit {
    private readonly appService = inject(AppService);

    ngOnInit(): void {
        this.appService.updateTitle('Meeting');
    }
}
