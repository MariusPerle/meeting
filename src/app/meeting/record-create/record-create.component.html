<dialog #leaveMeetingDialog>
  <div>
    Are you sure you want to leave the meeting?<br> All meeting data will be lost.
  </div>
  <div class="interactions">
    <button routerLink="/" aria-label="Leave" class="leave">Leave</button>
    <button (click)="leaveMeetingDialog.close()" aria-label="Cancel">Cancel</button>
  </div>
</dialog>
<div class="row">
  <div class="col navigation">
    <button (click)="goBack(leaveMeetingDialog)" aria-label="Go back">Go Back</button>
  </div>
  <div class="col updated-time">
    @if (timer) {
      {{ timer | async }}
    }
  </div>
</div>

<div class="meeting-title">
  <input id="title" type="text" [formControl]="title" placeholder="Enter meeting title (required to save)">
</div>

@for (member of team(); track member) {
  <div class="row team-member"
       [ngClass]="{
       even: $even,
       active: member.id === activeMember?.id,
       'not-attend': !meetingInfo[member.id].controls.attended.value,
       participated: meetingInfo[member.id].controls.participated.value
       }"
       [attr.data-test]="member.id + '-team-member'"
  >
    <div class="col">
      <div class="row">
        <div class="col">
          {{ member.name }}
        </div>
        <div class="col participated">
          @if (!meetingInfo[member.id].controls.participated.value) {
            <label [for]="'attended' + member.id">Attended</label>
            <input
              [id]="'attended' + member.id"
              [formControl]="meetingInfo[member.id].controls.attended"
              type="checkbox">
          }
          @if (meetingInfo[member.id].controls.attended.value) {
            <label [for]="'participated' + member.id">Participated</label>
            <input
              [id]="'participated' + member.id"
              [formControl]="meetingInfo[member.id].controls.participated"
              type="checkbox">
          }
        </div>
      </div>

    </div>
    <div class="col notes">
      <input
        [attr.aria-label]="member.name + 'note'"
        [formControl]="meetingInfo[member.id].controls.note"
        placeholder="Note"
        type="text">
    </div>
    <div class="col track-time">
      @if (member.randomPosition !== -1) {
        <div class="random-position">{{ member.randomPosition }}</div>
      }
      <div class="time">
        Time: {{ meetingInfo[member.id].value.time | date: 'mm:ss' }}
      </div>
      <button (click)="selectMember(member)" aria-label="Actives">Active</button>
    </div>
  </div>
} @empty {
  <p>No team members</p>
}
<div class="row">
  @if (!meetingInProgress) {
    <div class="col data-controls">
      <button (click)="resetMeeting()" aria-label="Reset Meeting">Reset Meeting</button>
      <button (click)="saveMeeting()"
              [disabled]="!title.valid"
              data-test="save-data"
              aria-label="Save Data">Save Data
      </button>
    </div>
  } @else {
    <div class="col end-meeting">
      <button (click)="endMeeting()" aria-label="Pause/Finish" data-test="end-meeting">Finish</button>
    </div>
  }
</div>

