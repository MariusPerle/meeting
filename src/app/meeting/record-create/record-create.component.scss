:host {
    dialog {
        .leave {
            margin-right: 1ch;
            background: red;
        }
    }

    .meeting-title {
        display: flex;
        justify-content: center;
        margin-bottom: 1rem;

        input {
            width: 33ch;
            text-align: center;
        }
    }

    .row {
        padding: 0.5rem 0;
        margin: 0;
    }

    .updated-time {
        display: flex;
        justify-content: flex-end;
        padding-right: 9ch;
    }

    .team-member {
        background: #eeeeee;

        &.even {
            background-color: #cacaca;
        }

        &.not-attend {
            background-color: rgba(204, 0, 0, 0.3);
        }

        &.participated {
            background-color: rgba(14, 197, 0, 0.3);
        }


        &.active {
            background-color: rgba(0, 47, 151, 0.3);
        }

        &:hover {
            background-color: rgba(0, 47, 151, 0.5);
        }

        .random-position {
            background: #824fce94;
            border-radius: 2ch;
            width: 3ch;
            height: 3ch;
            display: inline-flex;
            align-items: center;
            justify-content: center;
        }

        .interactions {
            display: flex;
            justify-content: flex-end;
        }

        .row {
            padding: 0;

            .participated {
                display: inline-flex;
                justify-content: center;
                align-items: center;

                label {
                    display: block;
                    font-size: small;
                    padding: 1ch;
                }
            }
        }

        .notes {
            min-width: 32ch;

            input[type="text"] {
                width: 100%;
                height: 100%;
            }
        }

        .track-time {
            display: flex;
            justify-content: flex-end;
            align-items: center;

            .time {
                height: 100%;
                line-height: 2rem;
            }

            button {
                margin-left: 1ch;
            }
        }
    }

    .data-controls {
        button {
            margin-right: 1ch;
        }
    }

    .end-meeting {
        display: flex;
        justify-content: flex-end;
    }
}