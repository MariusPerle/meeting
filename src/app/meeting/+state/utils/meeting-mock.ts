import { Record, RecordMember } from './meeting.interface';

export function recordFactory(
    overwrite: Pick<Record, 'id'> & Partial<Record>,
): Record {
    return {
        name: `Meeting`,
        date: new Date('2022-01-01'),
        members: [],
        ...overwrite,
    };
}

export function recordMemberFactory(
    overwrite: Pick<RecordMember, 'id'> & Partial<RecordMember>,
): RecordMember {
    return {
        time: 0,
        note: '',
        attended: true,
        participated: false,
        ...overwrite,
    };
}
