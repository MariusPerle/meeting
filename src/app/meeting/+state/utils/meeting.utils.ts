import { isMeeting, Record } from './meeting.interface';

export const meetingsStoreKey = 'meetings';

export function loadMeetingsFromLocalStorage(): Record[] {
    const localStorageValue = localStorage.getItem(meetingsStoreKey);
    if (!localStorageValue) {
        return [];
    }
    const meetings = JSON.parse(localStorageValue).map((meeting: Record) => ({
        ...meeting,
        date: new Date(meeting.date),
    }));

    if (!(Array.isArray(meetings) && meetings.every(isMeeting))) {
        return [];
    }
    return meetings;
}

export function saveToLocalStorage(history: Record[]) {
    localStorage.setItem(meetingsStoreKey, JSON.stringify(history));
}
