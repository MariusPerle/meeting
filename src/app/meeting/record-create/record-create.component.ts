import { Component, DestroyRef, inject, OnInit } from '@angular/core';
import { CommonModule } from '@angular/common';
import {
    FormControl,
    FormGroup,
    ReactiveFormsModule,
    Validators,
} from '@angular/forms';
import { Router, RouterLink } from '@angular/router';
import { selectMembers } from '../../+state/team/team.selectors';
import { TeamMember } from '../../+state/team/utils/team.interface';
import { interval, map, Observable, Subject, takeUntil } from 'rxjs';
import { Store } from '@ngrx/store';
import { AppStore } from '../../+state/appStore';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { saveRecord } from '../+state/meeting.actions';
import { RecordMember } from '../+state/utils/meeting.interface';

@Component({
    selector: 'meeting-recording-meeting',
    imports: [CommonModule, ReactiveFormsModule, RouterLink],
    templateUrl: './record-create.component.html',
    styleUrl: './record-create.component.scss',
})
export class RecordCreateComponent implements OnInit {
    private readonly store: Store<AppStore> = inject(Store);
    private readonly destroyRef = inject(DestroyRef);
    private router = inject(Router);

    team = this.store.selectSignal(selectMembers);
    title = new FormControl('', {
        nonNullable: true,
        validators: Validators.required,
    });
    meetingInfo: {
        [key: string]: FormGroup<{
            time: FormControl<number>;
            note: FormControl<string>;
            attended: FormControl<boolean>;
            participated: FormControl<boolean>;
        }>;
    } = {};
    activeMember?: TeamMember;
    startSpeakingTime = 0;

    timer?: Observable<string>;
    meetingInProgress = false;
    endTimer = new Subject<void>();

    ngOnInit(): void {
        this.team().forEach((member) => {
            this.meetingInfo[member.id] = new FormGroup({
                time: new FormControl(0, { nonNullable: true }),
                note: new FormControl('', { nonNullable: true }),
                attended: new FormControl(true, {
                    nonNullable: true,
                }),
                participated: new FormControl(false, {
                    nonNullable: true,
                }),
            });
        });
    }

    /**
     * set up timer to display current time mark mark current meeting as in progress
     */
    startMeeting() {
        if (this.timer) {
            return;
        }
        this.endTimer.next();
        this.meetingInProgress = true;
        this.timer = interval(1000).pipe(
            takeUntil(this.endTimer),
            takeUntilDestroyed(this.destroyRef),
            map((sec_num: number) =>
                new Date(sec_num * 1000 + 1000).toISOString().slice(11, 19),
            ),
        );
    }

    resetMeeting() {
        this.timer = undefined;
        Object.keys(this.meetingInfo).forEach((key) => {
            this.meetingInfo[key].reset();
        });
        this.title.reset();
    }

    /**
     * adds current time to active member
     */
    updateTime() {
        if (this.activeMember) {
            this.meetingInfo[this.activeMember.id].patchValue({
                time:
                    (this.meetingInfo[this.activeMember.id].value.time ?? 0) +
                    new Date().getTime() -
                    this.startSpeakingTime,
            });
        }
    }

    selectMember(member: TeamMember) {
        // start timer for display purposes
        this.startMeeting();

        this.updateTime();
        this.meetingInfo[member.id].patchValue({
            participated: true,
            attended: true,
        });

        // stop active member from being active
        if (this.activeMember?.id === member.id) {
            this.activeMember = undefined;
            this.startSpeakingTime = 0;
            return;
        }

        this.activeMember = member;
        this.startSpeakingTime = new Date().getTime();
    }

    endMeeting() {
        this.updateTime();

        this.activeMember = undefined;
        this.meetingInProgress = false;
        // end timer
        if (this.timer) {
            this.endTimer.next();
        }
    }

    saveMeeting() {
        const meetingMembers = this.team()
            .map((member) => {
                const meetingInfo = this.meetingInfo[member.id].value;
                if (!meetingInfo || !meetingInfo.attended) return undefined;
                const update: RecordMember = {
                    id: member.id,
                    time: meetingInfo.time ?? 0,
                    note: meetingInfo.note ?? '',
                    attended: meetingInfo.attended ?? false,
                    participated: meetingInfo.participated ?? false,
                };
                return update;
            })
            .filter((entry) => entry !== undefined) as RecordMember[];

        this.store.dispatch(
            saveRecord({
                meeting: { members: meetingMembers, name: this.title.value },
            }),
        );
        this.resetMeeting();
    }

    goBack(leaveMeetingDialog: HTMLDialogElement) {
        if (!this.meetingInProgress) {
            this.router.navigate(['/']);
        }
        leaveMeetingDialog.showModal();
    }
}
