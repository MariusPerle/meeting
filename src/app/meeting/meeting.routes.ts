import { Route } from '@angular/router';
import { RecordCreateComponent } from './record-create/record-create.component';
import { RecordListComponent } from './record-list/record-list.component';

export const meetingRoutes: Route[] = [
    { path: '', component: RecordCreateComponent },
    { path: 'list', component: RecordListComponent },
    { path: '**', redirectTo: '' },
];
