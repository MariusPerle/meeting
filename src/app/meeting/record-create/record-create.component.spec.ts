import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RecordCreateComponent } from './record-create.component';
import { provideMockStore } from '@ngrx/store/testing';
import { selectMembers } from '../../+state/team/team.selectors';
import { teamMemberFactory } from '../../+state/team/utils/team-mock';
import { RouterModule } from '@angular/router';
import { Store } from '@ngrx/store';
import { saveRecord } from '../+state/meeting.actions';

describe('RecordingMeetingComponent', () => {
    let component: RecordCreateComponent;
    let fixture: ComponentFixture<RecordCreateComponent>;
    const members = [
        teamMemberFactory({ id: '1' }),
        teamMemberFactory({ id: '2' }),
    ];

    beforeEach(async () => {
        jest.useFakeTimers();
        jest.setSystemTime(new Date('2023-01-01:00:00:00'));

        await TestBed.configureTestingModule({
            imports: [RecordCreateComponent, RouterModule.forRoot([])],
            providers: [
                provideMockStore({
                    selectors: [
                        {
                            selector: selectMembers,
                            value: members,
                        },
                    ],
                }),
            ],
        }).compileComponents();

        fixture = TestBed.createComponent(RecordCreateComponent);
        component = fixture.componentInstance;
        jest.spyOn(
            (component as unknown as { store: Store }).store,
            'dispatch',
        );

        fixture.detectChanges();
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render initial state', () => {
        expect(fixture).toMatchSnapshot();
    });

    it('should create record where all members are present and time was tracked', async () => {
        fixture.nativeElement
            .querySelector('[data-test="1-team-member"] .track-time button')
            .click();
        fixture.detectChanges();
        jest.setSystemTime(new Date('2023-01-01:00:01:00'));
        fixture.nativeElement
            .querySelector('[data-test="2-team-member"] .track-time button')
            .click();
        fixture.detectChanges();
        jest.setSystemTime(new Date('2023-01-01:00:02:40'));
        fixture.nativeElement
            .querySelector('[data-test="end-meeting"]')
            .click();
        fixture.detectChanges();
        expect(fixture).toMatchSnapshot();
        const title = fixture.nativeElement.querySelector('#title');
        title.value = 'Test';
        title.dispatchEvent(new Event('input'));
        fixture.detectChanges();
        fixture.nativeElement.querySelector('[data-test="save-data"]').click();
        expect(
            (component as unknown as { store: Store }).store.dispatch,
        ).toHaveBeenCalledWith(
            saveRecord({
                meeting: {
                    name: 'Test',
                    members: [
                        {
                            id: '1',
                            time: 60000,
                            note: '',
                            attended: true,
                            participated: true,
                        },
                        {
                            id: '2',
                            time: 100000,
                            note: '',
                            attended: true,
                            participated: true,
                        },
                    ],
                },
            }),
        );
    });

    it('should create record where one member was absent', async () => {
        fixture.nativeElement
            .querySelector('[data-test="1-team-member"] .track-time button')
            .click();
        fixture.detectChanges();
        jest.setSystemTime(new Date('2023-01-01:00:01:00'));
        fixture.nativeElement.querySelector('#attended2').click();
        const title = fixture.nativeElement.querySelector('#title');
        title.value = 'Test';
        title.dispatchEvent(new Event('input'));
        fixture.nativeElement
            .querySelector('[data-test="end-meeting"]')
            .click();
        fixture.detectChanges();
        fixture.nativeElement.querySelector('[data-test="save-data"]').click();
        expect(
            (component as unknown as { store: Store }).store.dispatch,
        ).toHaveBeenCalledWith(
            saveRecord({
                meeting: {
                    name: 'Test',
                    members: [
                        {
                            id: '1',
                            time: 60000,
                            note: '',
                            attended: true,
                            participated: true,
                        },
                    ],
                },
            }),
        );
    });
});
