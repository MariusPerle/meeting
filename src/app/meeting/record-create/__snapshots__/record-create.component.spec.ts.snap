// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordingMeetingComponent should create record where all members are present and time was tracked 1`] = `
<meeting-recording-meeting
  activeMember="undefined"
  destroyRef={[Function NodeInjectorDestroyRef]}
  endTimer={[Function Subject]}
  meetingInProgress="false"
  meetingInfo={[Function Object]}
  router={[Function _Router]}
  startSpeakingTime={[Function Number]}
  store={[Function _MockStore]}
  team={[Function Function]}
  timer={[Function Observable]}
  title={[Function FormControl2]}
>
  <dialog>
    <div>
       Are you sure you want to leave the meeting?
      <br />
       All meeting data will be lost. 
    </div>
    <div
      class="interactions"
    >
      <button
        aria-label="Leave"
        class="leave"
        routerlink="/"
        tabindex="0"
      >
        Leave
      </button>
      <button
        aria-label="Cancel"
      >
        Cancel
      </button>
    </div>
  </dialog><div
    class="row"
  >
    <div
      class="col navigation"
    >
      <button
        aria-label="Go back"
      >
        Go Back
      </button>
    </div>
    <div
      class="col updated-time"
    >
    </div>
  </div><div
    class="meeting-title"
  >
    <input
      class="ng-untouched ng-pristine ng-invalid"
      id="title"
      placeholder="Enter meeting title (required to save)"
      type="text"
    />
  </div><div
    class="row team-member even participated"
    data-test="1-team-member"
  >
    <div
      class="col"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
           Member 
        </div>
        <div
          class="col participated"
        >
          <label
            for="participated1"
          >
            Participated
          </label>
          <input
            class="ng-untouched ng-pristine ng-valid"
            id="participated1"
            type="checkbox"
          />
        </div>
      </div>
    </div>
    <div
      class="col notes"
    >
      <input
        aria-label="Membernote"
        class="ng-untouched ng-pristine ng-valid"
        placeholder="Note"
        type="text"
      />
    </div>
    <div
      class="col track-time"
    >
      <div
        class="random-position"
      >
      </div>
      <div
        class="time"
      >
         Time: 01:00 
      </div>
      <button
        aria-label="Actives"
      >
        Active
      </button>
    </div>
  </div><div
    class="row team-member participated"
    data-test="2-team-member"
  >
    <div
      class="col"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
           Member 
        </div>
        <div
          class="col participated"
        >
          <label
            for="participated2"
          >
            Participated
          </label>
          <input
            class="ng-untouched ng-pristine ng-valid"
            id="participated2"
            type="checkbox"
          />
        </div>
      </div>
    </div>
    <div
      class="col notes"
    >
      <input
        aria-label="Membernote"
        class="ng-untouched ng-pristine ng-valid"
        placeholder="Note"
        type="text"
      />
    </div>
    <div
      class="col track-time"
    >
      <div
        class="random-position"
      >
      </div>
      <div
        class="time"
      >
         Time: 01:40 
      </div>
      <button
        aria-label="Actives"
      >
        Active
      </button>
    </div>
  </div><div
    class="row"
  >
    <div
      class="col data-controls"
    >
      <button
        aria-label="Reset Meeting"
      >
        Reset Meeting
      </button>
      <button
        aria-label="Save Data"
        data-test="save-data"
        disabled=""
      >
        Save Data 
      </button>
    </div>
  </div>
</meeting-recording-meeting>
`;

exports[`RecordingMeetingComponent should render initial state 1`] = `
<meeting-recording-meeting
  destroyRef={[Function NodeInjectorDestroyRef]}
  endTimer={[Function Subject]}
  meetingInProgress="false"
  meetingInfo={[Function Object]}
  router={[Function _Router]}
  startSpeakingTime="0"
  store={[Function _MockStore]}
  team={[Function Function]}
  title={[Function FormControl2]}
>
  <dialog>
    <div>
       Are you sure you want to leave the meeting?
      <br />
       All meeting data will be lost. 
    </div>
    <div
      class="interactions"
    >
      <button
        aria-label="Leave"
        class="leave"
        routerlink="/"
        tabindex="0"
      >
        Leave
      </button>
      <button
        aria-label="Cancel"
      >
        Cancel
      </button>
    </div>
  </dialog><div
    class="row"
  >
    <div
      class="col navigation"
    >
      <button
        aria-label="Go back"
      >
        Go Back
      </button>
    </div>
    <div
      class="col updated-time"
    >
    </div>
  </div><div
    class="meeting-title"
  >
    <input
      class="ng-untouched ng-pristine ng-invalid"
      id="title"
      placeholder="Enter meeting title (required to save)"
      type="text"
    />
  </div><div
    class="row team-member even"
    data-test="1-team-member"
  >
    <div
      class="col"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
           Member 
        </div>
        <div
          class="col participated"
        >
          <label
            for="attended1"
          >
            Attended
          </label>
          <input
            class="ng-untouched ng-pristine ng-valid"
            id="attended1"
            type="checkbox"
          />
          <label
            for="participated1"
          >
            Participated
          </label>
          <input
            class="ng-untouched ng-pristine ng-valid"
            id="participated1"
            type="checkbox"
          />
        </div>
      </div>
    </div>
    <div
      class="col notes"
    >
      <input
        aria-label="Membernote"
        class="ng-untouched ng-pristine ng-valid"
        placeholder="Note"
        type="text"
      />
    </div>
    <div
      class="col track-time"
    >
      <div
        class="random-position"
      >
      </div>
      <div
        class="time"
      >
         Time: 00:00 
      </div>
      <button
        aria-label="Actives"
      >
        Active
      </button>
    </div>
  </div><div
    class="row team-member"
    data-test="2-team-member"
  >
    <div
      class="col"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
           Member 
        </div>
        <div
          class="col participated"
        >
          <label
            for="attended2"
          >
            Attended
          </label>
          <input
            class="ng-untouched ng-pristine ng-valid"
            id="attended2"
            type="checkbox"
          />
          <label
            for="participated2"
          >
            Participated
          </label>
          <input
            class="ng-untouched ng-pristine ng-valid"
            id="participated2"
            type="checkbox"
          />
        </div>
      </div>
    </div>
    <div
      class="col notes"
    >
      <input
        aria-label="Membernote"
        class="ng-untouched ng-pristine ng-valid"
        placeholder="Note"
        type="text"
      />
    </div>
    <div
      class="col track-time"
    >
      <div
        class="random-position"
      >
      </div>
      <div
        class="time"
      >
         Time: 00:00 
      </div>
      <button
        aria-label="Actives"
      >
        Active
      </button>
    </div>
  </div><div
    class="row"
  >
    <div
      class="col data-controls"
    >
      <button
        aria-label="Reset Meeting"
      >
        Reset Meeting
      </button>
      <button
        aria-label="Save Data"
        data-test="save-data"
        disabled=""
      >
        Save Data 
      </button>
    </div>
  </div>
</meeting-recording-meeting>
`;
